### PR TITLE
[webui][ci] Initial tests for model Comment

### PR DESCRIPTION
**Pull request by eduardoj**

### CODE REVIEW

**Comment by Ana06** [comment]
I think that the commit message is not very clear. If what you meant is that you haven't tested all the method what about something like

```
[ci] Create test for Comment model

Not all the methods of the model have been tested.
```

Also, does the [webui] tag make sense here?

**Comment by Ana06** [comment]
You could also test the `belong_to` relations. For example for the first one:

` it { should belong_to(:bs_request).inverse_of(:comments) }`

File: src/api/spec/factories/comment.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+  factory :comment do
+    body { Faker::Lorem.paragraph }
+    type { Faker::Lorem.word }
```
Comment (Ana06) [inline src/api/spec/factories/comment.rb]: You shouldn't use a random word as a `type` as it is reserved for storing the the inheritance class. Because of the same reason and taking into account that it is not possible to create a `comment `without type (because of the presence validation) it doesn't make sense to test `comment.` You should test the comment children. I think that for some function it is enough with testing it for one child (for example comment_package), although for some of them it may be needed to test it for more.

Also, I added a validation for that in https://github.com/openSUSE/open-build-service/pull/2252 and this should fail in the tests after it is merged (it already fails outside the tests).

File: src/api/spec/models/comment_spec.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
+require "rails_helper"
+
+RSpec.describe Comment do
+  let!(:comment) { create(:comment) }
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: You don't need the `!`

In case that you need the object to be created (like happens in the `blank_or_destroy` test) you can do

``` Ruby
  describe "blank_or_destroy" do
    before do
      comment
    end
```
That way you don't create it for all tests.

File: src/api/spec/models/comment_spec.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+            expect(comment_element.text).to match(/^#<Nokogiri::XML::Builder::NodeBuilder:0x\h+>$/)
+          end
+        end
+      end
+    end
+  end
+
+  describe "blank_or_destroy" do
+    context "without children" do
```
Comment (eduardoj) [inline src/api/spec/models/comment_spec.rb]: Added.

**Comment by hennevogel** [comment]
@eduardoj what's the status of this one?

**Comment by eduardoj** [comment]
@hennevogel: @Ana06, @bgeuken and me have been working on it, and still are. We are facing problems when we execute the whole test suite, but not when we execute the test alone. We have created a branch with some changes: https://github.com/eduardoj/open-build-service/tree/test-comment

**Comment by Ana06** [comment]
Yes, how the database is cleaned after every factory is wrong as it is deleting the users in the seeds and this needs to be solved before writing this test. :wink:

**Comment by hennevogel** [comment]
So I guess you should rebase now :-)

File: src/api/spec/models/comment_spec.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        expect { comment.blank_or_destroy }.to change { Comment.count }.by(-1)
+      end
+    end
+
+    context "with children" do
+      before do
+        create(comment_type, parent: comment)
+      end
+
+      it 'should be destroyed' do
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: This should be `it 'shouldn't be destroyed' do`

File: src/api/spec/models/comment_spec.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        expect { comment.blank_or_destroy }.to_not change { Comment.count }
+
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: why is it this blank line here?

File: src/api/spec/models/comment_spec.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        expect(comment.body).to eq 'This comment has been deleted'
+        expect(comment.user).to eq User.find_nobody!
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: I would test check the user login instead of the whole User to ensure that the find_nobody method is working properly:

`expect(comment.user.login).to eq '_nobody_'`

File: src/api/spec/models/comment_spec.rb [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+  describe "Comment validations" do
+    it { is_expected.to validate_presence_of(:body) }
+    it { is_expected.to validate_presence_of(:type) }
+    it { is_expected.to validate_presence_of(:user) }
+  end
+
+  describe "to_xml" do
+    context "without parent" do
+      let(:builder) {
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: This is still strange. Why not doing 
``` Ruby
let(:builder) { builder = Nokogiri::XML::Builder.new }
```
and then

``` Ruby
before do
  comment.to_xml(builder)
end
```

that should be enough

File: src/api/spec/models/comment_spec.rb [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        comment.to_xml(builder)
+        builder
+      }
+
+      context "returns xml" do
+        context "comment element" do
+          let(:comment_element) { builder.doc.css('comment') }
+
+          it "comment attributes and content" do
+            expect(comment_element.attribute('id').value).to match(/^\d+$/)
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: why don't you check the exactly value?

File: src/api/spec/models/comment_spec.rb [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    it { expect(build(comment_type)).to be_valid }
+  end
+
+  describe "Comment associations" do
+    it { is_expected.to belong_to(:bs_request).inverse_of(:comments) }
+    it { is_expected.to belong_to(:project).inverse_of(:comments) }
+    it { is_expected.to belong_to(:package).inverse_of(:comments) }
+    it { is_expected.to belong_to(:user).inverse_of(:comments) }
+
+    it { is_expected.to have_many(:children) }
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: why not 

`it { is_expected.to have_many(:children).dependent(:destroy).class_name('Comment').with_foreign_key('parent_id') }`

http://www.rubydoc.info/github/thoughtbot/shoulda-matchers/Shoulda%2FMatchers%2FActiveRecord%3Ahave_many

**Comment by Ana06** [comment]
Please delete the last commit, shouldnt be there

**Comment by Ana06** [comment]
Squash https://github.com/openSUSE/open-build-service/pull/2247/commits/8b5002adb6421aa3d20d905d43764c187ea6fcea into https://github.com/openSUSE/open-build-service/pull/2247/commits/28d1740f27246ecfd756d8fa8888707db6f5cfba :wink:

File: src/api/spec/models/comment_spec.rb [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        comment.to_xml(builder)
+      end
+
+      context "returns xml" do
+        context "comment element" do
+          let(:comment_element) { builder.doc.css('comment') }
+
+          it "comment attributes and content" do
+            expect(comment_element.attribute('id').value).to eq(comment.id.to_s)
+            expect(comment_element.attribute('when').value).to match(/^\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d UTC$/)
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: why don't you check the exactly value?

File: src/api/spec/models/comment_spec.rb [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+            expect(comment_element.attribute('who').value).to match(/^user_\d+$/)
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: why don't you check the exactly value?

File: src/api/spec/models/comment_spec.rb [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+            expect(comment_element.text).to match(/^#<Nokogiri::XML::Builder::NodeBuilder:0x\h+>$/)
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: why don't you check the exactly value?

File: src/api/spec/models/comment_spec.rb [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+          end
+        end
+      end
+    end
+
+    context "with parent" do
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: This tests is exactly the same that the one above but with one more attribute. Why have you left it empty? I would prefer to have the `with parent` case as it is "more general".

Comment (eduardoj) [inline src/api/spec/models/comment_spec.rb]: Done.

File: src/api/app/models/comment.rb [hunk 14]
```diff
@@ -60,8 +60,8 @@ def to_xml(builder)
     attrs = { who: user, when: created_at, id: id }
     attrs[:parent] = parent_id if parent_id
 
-    builder.comment_(attrs) do
-      builder.text(body)
+    builder.comment_(attrs) do |xml|
```
Comment (Ana06) [inline src/api/app/models/comment.rb]: Why did you change that?

File: src/api/spec/models/comment_spec.rb [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+      before do
+        comment.to_xml(builder)
+      end
+
+      context "returns xml" do
+        context "comment element" do
+          let(:comment_element) { builder.doc.css('comment') }
+
+          it "comment attributes and content" do
+            expect(comment_element.attribute('id').value).to eq(comment.id.to_s)
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: Isn't it possible to do that using the result of the `.to_xml` ?

File: src/api/spec/models/comment_spec.rb [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    let(:builder) { builder = Nokogiri::XML::Builder.new }
+
+    context "without parent" do
+      before do
+        comment.to_xml(builder)
+      end
+
+      context "returns xml" do
+        context "comment element" do
+          let(:comment_element) { builder.doc.css('comment') }
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: `let(:comment_element) { builder.doc.css('comment') }` is also several time, please move it up so you don't need to have it twice.

File: src/api/spec/models/comment_spec.rb [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    it { is_expected.to have_many(:children).dependent(:destroy).class_name('Comment').with_foreign_key('parent_id') }
+  end
+
+  describe "Comment validations" do
+    it { is_expected.to validate_presence_of(:body) }
+    it { is_expected.to validate_presence_of(:type) }
+    it { is_expected.to validate_presence_of(:user) }
+  end
+
+  describe "to_xml" do
```
Comment (Ana06) [inline src/api/spec/models/comment_spec.rb]: I am not pretty sure if there is a better way to test this function. @bgeuken can you take a look?

File: src/api/spec/models/comment_spec.rb [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    let(:builder) { builder = Nokogiri::XML::Builder.new }
```
Comment (ChrisBr) [inline src/api/spec/models/comment_spec.rb]: The assignment is not necessary, is it?

File: src/api/spec/models/comment_spec.rb [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+        expect { comment.blank_or_destroy }.to change { Comment.count }.by(-1)
+      end
+    end
+
+    context "with children" do
+      before do
+        create(comment_factory, parent: comment)
+      end
+
+      it 'shouldn\'t be destroyed' do
```
Comment (ChrisBr) [inline src/api/spec/models/comment_spec.rb]: I would prefer double quotes instead of escaping the single quote!
"shouldn't be destroyed"

**Comment by ChrisBr** [comment]
@eduardoj Is this still dependent on #2297 or ready to merge?

**Comment by Ana06** [comment]
@ChrisBr it doesn't depends on #2297

**Comment by bgeuken** [comment]
Looks good. Let's ship this!
